modules/javafx.graphics/src/main/java/javafx/scene/Scene.java

```diff
@@ -589,6 +589,29 @@ private void doCSSPass() {
             sceneRoot.clearDirty(com.sun.javafx.scene.DirtyBits.NODE_CSS);
             sceneRoot.processCSS();
         }
+
+        if (!clearInitialCssStateNodes.isEmpty()) {
+            for (var node : clearInitialCssStateNodes) {
+                node.clearInitialCssStateFlag();
+            }
+
+            clearInitialCssStateNodes.clear();
+        }
+    }
+
+    /**
+     * A list of nodes that have expressed their interest to be notified when the next CSS pass
+     * is completed. Nodes will use this event to determine whether they are in their initial
+     * CSS state (see {@link Node#initialCssState}.
+     */
+    private final Set<Node> clearInitialCssStateNodes = Collections.newSetFromMap(new IdentityHashMap<>());
```

Comment:
why not use a normal `HashSet`? What advantage has this approach?

Reply:
`HashSet` uses `Object.equals`, which can be overridden by user code, and this would break the logic. It's the _instance_ that wants to be notified.

Reply:
Good point, I never ever did that for `Node`s (and I don't know why I would need to), but you are right, it is indeed possible and therefore a possible scenario.

Reply:
I think this is overly cautious.  Overriding equals on a class that did not implement equals in a hierarchy you don't control  is not a very reasonable scenario.  You will not be able to call `super.equals` with reasonable results if the hierarchy did not implement it in the first place.  This is because there may be private data that you can't access for your equality comparison (and `Node` has lots of that).  We also use `WeakHashMap`s in several areas already with subtypes of `Node`s as keys (I found `Region`, `Parent` and `TreeView` being used as keys), so this kind of override should probably be documented as being unsupported on `Node`.

Reply:
Makes sense. I think we should make `Node.equals()` final in this case.